### ISSUE TITLE
temples aren't deleted

### ISSUE THREAD

**Issue by pgierz** [Try2Code/cdo-bindings#18]
It appears as if the `__del__` method of `MyTempFile` is never called, leading to the `/tmp` folder on my system to be filled. A reboot would solve this problem, but how can this be done on computers that are rarely rebooted (e.g. servers)

**Comment by pgierz** [comment]
I've found part of the problem: python's `tempfile.NamedTemporaryFile` are only deleted if the files are explicitly closed.

**Comment by Try2Code** [comment]
hi! @pgierz 

sorry for the huge delay. are you willing to run some test for me? this kind of behavior is always hard to test...

**Comment by pgierz** [comment]
Sure, no problem! I'm not sure yet if I'll get to it today, but if you outline what information you need, I'll see what I can do.

Cheers,
Paul

**Comment by pgierz** [comment]
Haha alright :-) 

I'll redo one of the examples I had our interns do a few weeks ago. Is there a way to explicitly turn on verbose output? That could make our debugging efforts easier.

**Comment by pgierz** [comment]
Ok, here's a quick and dirty way to demonstrate the problem. In the first case, we get desired behaviour. The python program exits, and the temporary file is gone.
```sh
$ ipython
Python 3.6.3 |Anaconda custom (64-bit)| (default, Oct 13 2017, 12:02:49) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.1.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import cdo

In [2]: cdo.Cdo()
Out[2]: <cdo.Cdo at 0x7f1e05c2f400>

In [3]: CDO = cdo.Cdo()

In [4]: blah = CDO.fldmean(input="/scratch/simulation_database/incoming/Eem120-S2/output/Eem120-S2_echam5_main_mm_260001.nc")

In [5]: blah
Out[5]: '/tmp/cdoPy01hj7uva'

In [6]: exit()
$ ls /tmp/cdoPy01hj7uva
ls: cannot access /tmp/cdoPy01hj7uva: No such file or directory
```

In the second case, *something* (could be an SSH Timeout, an incorrectly exited Jupyter Notebook, etc) kills your python job:

```shell
$ ipython
Python 3.6.3 |Anaconda custom (64-bit)| (default, Oct 13 2017, 12:02:49) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.1.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import cdo

In [2]: CDO = cdo.Cdo()

In [3]: blah = CDO.fldmean(input="/scratch/simulation_database/incoming/Eem120-S2/output/Eem120-S2_echam5_main_mm_260001.nc")

In [4]: blah
Out[4]: '/tmp/cdoPyiz3pm_fj'
C-z
$ ps
   PID TTY          TIME CMD
 85224 pts/0    00:00:00 bash
 90565 pts/0    00:00:01 ipython
 90871 pts/0    00:00:00 ps
$ kill -9 90565
$ ls /tmp/cdoPyiz3pm_fj
/tmp/cdoPyiz3pm_fj
```

As I'm writing this, I realise this might actually be "normal" behaviour, because of course python can't clean up it's temporary files if it doesn't get an explicit exit command and is ended in some other way. Unless there's some way to catch that, we could close this and I would just need to enforce some stricter policies about how to end your programs...

**Comment by Try2Code** [comment]
when you run a script, you can add ```DEBUG=1``` on the command line or tell the cdo objects to be chatty with 

```python
cdo = Cdo()
cdo.debug = True
```

but the point is how to create the crash ....

**Comment by Try2Code** [comment]
aa, ok. if you work interactively. things cannot be removed. I added a try...finally loop around. that might help with that.  could you check the latest revision ?

**Comment by pgierz** [comment]
Ok, I'll try it out

Another idea I just had: I assume that the cdo module always tries to write files into the `/tmp/` directory. Would it be difficult to tell the constructor:
```
cdo = Cdo()
``` 
to check for the presence of temporary cdo files that belong to your user, and if so, erase them? Potentially with a flag, e.g. `delete_old_temp_files=True`. Otherwise, if you initialise multiple `Cdo()` in your script, it might delete files you're still trying to work on.

I could try this out in a branch and let you know if it works...

**Comment by Try2Code** [comment]
hm ... I'd like to have cdo removed everything it created. I guess, this needs a signal handler. I didnt have interactive use in mind when I started to write cdo.py, but many ppl use it that way. I had a similar request last week ...

**Comment by pgierz** [comment]
How about this: https://github.com/pgierz/cdo-bindings/tree/clean_up_tmp

You can give `cdo = Cdo(clean_tmp=True)`, and it'll look in the `/tmp` directory for any files that belong to you, and have the pattern `cdoPy` and erase them.

**Comment by pgierz** [comment]
By the way, the `try..finally` block didn't seem to help, the temporary files are still there if my python job is killed.

**Comment by Try2Code** [comment]
could you give the current revision another try, please! I added a signal handler and removed the try...finally loop.

your idea with user-requested deletion of tempfiles is applicable to single-user setups, but on servers it is different.  temp directories area global, so I would risk deletion of other peoples files. Access rights might help preventing this, but my first goal is a clean exit. so that I don't have to rely on system setups.

**Comment by pgierz** [comment]
I had added a check that the user trying to delete the file also owns the file for that very reason, but admittedly I can't really test it without having someone else's files in the `/tmp` folder ;-)

I'll pull the new revision and let you know!

**Comment by Try2Code** [comment]
I think that my last version *does* remove files when you kill an interactive session like in your example. Could you x-check it, please?

**Comment by Try2Code** [comment]
right, catching SIGKILL is not possible, so in that case tempfiles cannot be handled. jupyter notebooks might be more complex because of different object handling on the server side.

Detecting the left-overs might be the best strategy

**Comment by Try2Code** [comment]
I will be off for 3 weeks, so please keep you fork alive - I will merge it when I am back.

thx for your help!

**Comment by pgierz** [comment]
No problem, I'll keep my fork around!

Enjoy your holiday!

**Comment by Try2Code** [comment]
release 1.4.0 has much of the code from your branch. could u give a shot?

thx
